scipy/stats/stats.py

```diff
@@ -8121,6 +8121,281 @@ def brunnermunzel(x, y, alternative="two-sided", distribution="t",
+    bd = distributions.binom(n, quantile)
+
+    # the lower bound is the last index before the invert survival
+    # function value for the target confidence level
+    lb = bd.isf(confidence) - 1
+
+    if lb < 0:  # isf returns -1 if there are no matching index
+        return None
+    else:
+        return int(lb)
+
+
+def confint_quantile(x, quantile, confidence, type='one-sided'):
+    r"""Compute non-parametric confidence intervals for any quantile.
+
+    This function implements a non-parametric approach to compute confidence
+    intervals for quantiles. The approach is attributed to Thompson [1]_ and
+    later proven to be applicable to any set of i.i.d. samples [2]_. The
+    computation is based on the observation that the probability of a quantile
+    :math:`q` to be larger than any sample :math:`x_m (1\leq m \leq N)` can be
+    computed as
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = 1 - \sum_{k=0}^{m-1} \binom{N}{k}
+        q^k(1-q)^{N-k}
+
+    Furthermore, these probabilities are symmetric, which allows to compute both
+    upper and lower bounds from the same computation:
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = \mathbb{P}(x_{N-m+1} \geq 1-q).
+
+    The function computes confidence intervals for a given quantile and
+    confidence level based on `x`, which is either a set of samples
+    (one-dimensional array_like) or the number of samples available. The
+    confidence intervals are valid if and only if the samples are i.i.d.
+
+    Both one-sided and two-sided confidence intervals can be obtained (default
+    is one-sided). The function returns two values: either the bounds for the
+    two one-sided confidence intervals, or the lower and upper bounds of a
+    two-sided confidence interval.
+
+    If `x` is the set of samples, the return values are the sample values
+    corresponding with the bounds of the confidence interval. If `x` is an
+    integer, the function considers `x` as the sample size and returns the
+    indexes of the bounds for the confidence interval for this sample size. Note
+    that this is possible because the numerical values of the data are
+    irrelevant for this computing confidence interval using this method.
+
+    A similar function is available in the QuantileNPCI R package [3]_. The
+    foundation is the same, but it computes the confidence interval bounds by
+    doing interpolations between the sample values, whereas this function uses
+    only sample values as bounds. Thus, `confint_quantile` returns slightly more
+    conservative intervals (i.e., larger).
+
+    Parameters
+    ----------
+    x : array_like or int Array of samples, should be one-dimensional. If
+        integer, taken as the number of samples available (strictly positive)
+    quantile : float The quantile for which we want to compute the confidence 
+        interval. Must be strictly between 0 and 1. 
+    confidence : float The desired confidence level of the confidence interval. 
```

Comment:
Consider a default value of 0.95. I think there are other places in `scipy.stats` where this is the default. 
Also consider the name `confidence_level`. See [`bootstrap`](http://scipy.github.io/devdocs/reference/generated/scipy.stats.bootstrap.html#scipy.stats.bootstrap) and [`binomtest`](http://scipy.github.io/devdocs/reference/generated/scipy.stats.binomtest.html#scipy.stats.binomtest).